nomad/structs/structs_test.go

```diff
@@ -1665,6 +1665,55 @@ func TestTask_Validate_LogConfig(t *testing.T) {
 	}
 }
 
+func TestTask_Validate_CSIPluginConfig(t *testing.T) {
+	table := []struct {
+		name        string
+		pc          *TaskCSIPluginConfig
+		expectedErr string
+	}{
+		{
+			name: "no errors when not specified",
+			pc:   nil,
+		},
+		{
+			name:        "requires non-empty plugin id",
+			pc:          &TaskCSIPluginConfig{},
+			expectedErr: "CSIPluginConfig must have a non-empty PluginID",
+		},
+		{
+			name: "requires valid plugin type",
+			pc: &TaskCSIPluginConfig{
+				ID:   "com.hashicorp.csi",
+				Type: "nonsense",
+			},
+			expectedErr: "CSIPluginConfig PluginType must be one of 'node', 'controller', or 'monolith', got: \"nonsense\"",
+		},
+	}
+
+	for _, tt := range table {
+		t.Run(tt.name, func(t *testing.T) {
+			task := &Task{
+				CSIPluginConfig: tt.pc,
+			}
+			ephemeralDisk := &EphemeralDisk{
+				SizeMB: 1,
+			}
+
+			err := task.Validate(ephemeralDisk, JobTypeService, nil)
+			mErr := err.(*multierror.Error)
+			if tt.expectedErr != "" {
+				if !strings.Contains(mErr.Errors[4].Error(), tt.expectedErr) {
```

Comment:
So I'm somewhat conflicted about this, because we could generally abstract these tests with something like:

```
diff --git a/nomad/structs/structs_test.go b/nomad/structs/structs_test.go
index bb10058e8..353d3b0dd 100644
--- a/nomad/structs/structs_test.go
+++ b/nomad/structs/structs_test.go
@@ -16,6 +16,18 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func requireMultierrContains(t *testing.T, mErr *multierror.Error, expected string) {
+	t.Helper()
+
+	for _, err := range mErr.Errors {
+		if strings.Contains(err.Error(), expected) {
+			return
+		}
+	}
+
+	t.Fatalf("Expected multierror to contain: %v, instead had: %v", expected, mErr.Errors)
+}
+
 func TestJob_Validate(t *testing.T) {
 	j := &Job{}
 	err := j.Validate()
@@ -1648,8 +1660,8 @@ func TestTask_Validate_ConnectProxyKind(t *testing.T) {
 			}
 		})
 	}
-
 }
+
 func TestTask_Validate_LogConfig(t *testing.T) {
 	task := &Task{
 		LogConfig: DefaultLogConfig(),
@@ -1702,13 +1714,7 @@ func TestTask_Validate_CSIPluginConfig(t *testing.T) {
 			err := task.Validate(ephemeralDisk, JobTypeService, nil)
 			mErr := err.(*multierror.Error)
 			if tt.expectedErr != "" {
-				if !strings.Contains(mErr.Errors[4].Error(), tt.expectedErr) {
-					t.Fatalf("err: %s", err)
-				}
-			} else {
-				if len(mErr.Errors) != 4 {
-					t.Fatalf("unexpected err: %s", mErr.Errors[4])
-				}
+				requireMultierrContains(t, mErr, tt.expectedErr)
 			}
 		})
 	}
```

However, I think this would first require some level of validation that every error mode is well tested, because we probably gain some safety right now with various tests accidentally covering other cases.

I'm tempted to file a follow up issue for cleaning up the validation testing across the board tho.

Reply:
> I'm tempted to file a follow up issue for cleaning up the validation testing across the board tho.

Yeah, good call. Let's not clutter up this PR with that.

Reply:
#6844